### PR TITLE
 Update Mochiweb and add SameSite support to auth cookie

### PR DESCRIPTION
**Pull request by willholley**
## Overview

Adds a new configuration field, `couch_httpd_auth.same_site` which sets the `SameSite` attribute of the CouchDB auth cookie. If no value is set (the default), no `SameSite` attribute is added.

This also updates Mochiweb to v2.20 as this is required for `SameSite` support.

## Testing recommendations

Set the `samesite: strict` value in default.ini and verify that the SameSite attribute is detected by a browser (e.g. Chrome dev tools).

## Related Issues or Pull Requests

 * Refs #2221
 * Documentation PR: https://github.com/apache/couchdb-documentation/pull/473

## Checklist

- [ ] Code is written and works correctly
- [x] Changes are covered by tests
- [x] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [x] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

### CODE REVIEW

**Comment by wohali** [comment]
@willholley Thank you for this! I'm adding this to 3.0.0 for tracking purposes.

Do you think it would be wise (or unwise) to change the default to be `samesite = strict`?

File: rel/overlay/etc/default.ini [hunk 1]
```diff
@@ -242,9 +242,11 @@ iterations = 10 ; iterations for password hashing
 ; proxy_use_secret = false
 ; comma-separated list of public fields, 404 if empty
 ; public_fields =
-; secret =
```
Comment (rnewson) [inline rel/overlay/etc/default.ini]: lots of unrelated ws changes, fix pls

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+    "python.pythonPath": ".venv/bin/python3"
```
[the remark this answers is not in the document]
Comment (wohali) [reply .vscode/settings.json]: Good catch, please add `.vscode` to `.gitignore` and remove from the PR

**Comment by willholley** [comment]
thanks both - unrelated whitespace / vscode changes now removed

**Comment by willholley** [comment]
> Do you think it would be wise (or unwise) to change the default to be `samesite = strict`?

@wohali I'm wary of this in case it breaks clients that can't handle the `SameSite` attribute. I think for the majority of users, `strict` would be the better default from a security POV, but we probably want to at least test replication against older CouchDB, PouchDB etc before making it the default. Excluding `SameSite` should be fine - it's only users using CORS that will need to set it to `None` to keep things working.

**Comment by wohali** [comment]
Cool, I poked Bob who is going to re-read the cookie spec at the very least.

**Comment by YetAnotherEye** [comment]
Hi there,
would you consider to also add secure = true to the cookies?
I'm still getting
`A cookie associated with a resource at http://app.mitas-app.de/ was set with `SameSite=None` but without `Secure`. It has been blocked, as Chrome now only delivers cookies marked `SameSite=None` if they are also marked `Secure`. You can review cookies in developer tools under Application>Storage>Cookies and see more details at https://www.chromestatus.com/feature/5633521622188032.`

Thereby I'm not able to connect using our hybrid app served from localhost and I'm afraid that it also won't work in webviews. Any workarounds?
Thank you in advance.

**Comment by wohali** [comment]
@YetAnotherEye Does setting SameSite to `lax` or `strict` not work for you?

@willholley per https://web.dev/samesite-cookies-explained/ also, the user is right:

> Cookies with SameSite=None must also specify Secure, meaning they require a secure context.

We probably need a new PR for this.

**Comment by YetAnotherEye** [comment]
@wohali Hi, no, it doesn't. I seem to be one of the lucky persons that got the incremental update early (Version 80.0.3987.149 (Official) (64-Bit) on Win10.
As specified, Chrome doesn't save cross-site cookies set to lax or strict when using CouchDB from localhost. For development purposes I could use this workaround: https://support.google.com/chrome/thread/26291731?hl=en
But for now I'm simply using Chrome Canary, which still works.

But afaik, Cordova apps also use localhost as request host, what could/will become a problem in future. I tried to adapt the cookie using using apaches mod_rewrite that handles requests to the docker container, where the CouchDB is running - but failed. It would really make my day, if configuring that could be done natively in Couch-Settings.

**Comment by willholley** [comment]
@wohali @YetAnotherEye CouchDB [should automatically](https://github.com/apache/couchdb/blob/2b95500ac57d831ed4ac7abfc9250c739179b230/src/couch/src/couch_httpd_auth.erl#L463) add the `Secure` directive to the cookie so long as CouchDB is accessed over `https` (required for `SameSite=None` to work).

**Comment by YetAnotherEye** [comment]
@willholley Ok, i see it. So I assume that we have to do the internal communication between Apache and Couch via HTTPS. I'll check that out.
Thank you so far ;)

**Comment by YetAnotherEye** [comment]
Small update:
Setting ProxyPass and ProxyPassReverse to `https://localhost:6984` didn't work. Maybe because we don't have a certificate in the docker? Not my main area of competence ;)

I talked to our admin, who said, that we use 
`RequestHeader           set "X-Forwarded-Proto" expr=%{REQUEST_SCHEME}`
in our location config, and that this should tell the Couch-Container that the request was done via Https. 

To add the secure flag I now, according to
https://stackoverflow.com/questions/24129201/add-secure-and-httponly-flags-to-every-set-cookie-response-in-apache-httpd
used `Header edit Set-Cookie ^(.*)$ $1;HttpOnly;Secure` in Apache's location config.

I know, that's a security issue nonetheless the cookie with SameSite=none and Secure=true comes from CouchDB or the ReverseProxy, but for now we have a solution for our usecase.

Thanks for the hints and respect for the really good job you all are doing.

**Comment by thebigh2014** [comment]
Setting the Secure and SameSite=None attributes can break older browsers - especially Chrome (for my project, 66 and older break) - so this behaviour really needs to be adaptable to different UserAgents.  IOW, it should NOT be applied universally.

On my Apache reverse proxy, I only add both attributes if the UserAgent contains Chrome/80 or newer.  If other, non-Chrome browsers, begin requiring it then I will update the condition then but not before.
